File: sdks/typescript/test/client.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { AuthzClient } from "../src/client.js";
import { ToolGuard, MCPGuard, startAgentSession } from "../src/guard.js";
import { AuthzServiceError, PermissionDeniedError } from "../src/errors.js";

/**
 * Minimal mock of the global fetch function. Each invocation increments a
 * counter so tests can assert on retry / cache behavior, and the handler
 * decides what to return based on the request URL + body.
 */
function makeMockFetch(
  handler: (url: string, init?: RequestInit) => { status: number; body: unknown },
): { fetch: typeof fetch; calls: () => number } {
  let calls = 0;
  const mock: typeof fetch = async (input, init) => {
    calls++;
    const url = typeof input === "string" ? input : input.toString();
    const { status, body } = handler(url, init);
    return new Response(JSON.stringify(body), {
      status,
      headers: { "content-type": "application/json" },
    });
  };
  return { fetch: mock, calls: () => calls };
}

test("authorize returns allow result", async () => {
  const { fetch } = makeMockFetch(() => ({
    status: 200,
    body: {
      allowed: true,
      decision: "allow",
      reason: "permission_granted",
      required_permission: "docs.read",
      matched_permissions: ["docs.read"],
    },
  }));
  const client = new AuthzClient({ baseUrl: "http://x", apiKey: "k", fetchFn: fetch });
  const result = await client.authorize({
    tenantId: "t",
    applicationId: "a",
    subject: { type: "user", userId: "u" },
    resource: "docs",
    action: "read",
  });
  assert.equal(result.allowed, true);
  assert.equal(result.requiredPermission, "docs.read");
});

test("require throws PermissionDeniedError when denied", async () => {
  const { fetch } = makeMockFetch(() => ({
    status: 200,
    body: {
      allowed: false,
      decision: "deny",
      reason: "missing_permission",
      required_permission: "docs.write",
      matched_permissions: [],
    },
  }));
  const client = new AuthzClient({ baseUrl: "http://x", fetchFn: fetch });
  await assert.rejects(
    () =>
      client.require({
        tenantId: "t",
        applicationId: "a",
        subject: { type: "user", userId: "u" },
        resource: "docs",
        action: "write",
      }),
    (err: unknown) => err instanceof PermissionDeniedError,
  );
});

test("getEffectivePermissions caches by subject", async () => {
  const { fetch, calls } = makeMockFetch(() => ({
    status: 200,
    body: { permissions: ["docs.read", "tools.gmail.send"] },
  }));
  const client = new AuthzClient({
    baseUrl: "http://x",
    fetchFn: fetch,
    cacheTtlMs: 60_000,
  });
  for (let i = 0; i < 3; i++) {
    const perms = await client.getEffectivePermissions({
      tenantId: "t",
      applicationId: "a",
      subject: { type: "user", userId: "u" },
    });
    assert.ok(perms.has("docs.read"));
  }
  assert.equal(calls(), 1, "expected single round-trip thanks to LRU cache");
});

test("post retries on 5xx", async () => {
  let attempt = 0;
  const fetch: typeof globalThis.fetch = async () => {
    attempt++;
    if (attempt === 1) {
      return new Response("boom", { status: 500 });
    }
    return new Response(
      JSON.stringify({ allowed: true, decision: "allow", reason: "ok", required_permission: "x.y", matched_permissions: [] }),
      { status: 200, headers: { "content-type": "application/json" } },
    );
  };
  const client = new AuthzClient({
    baseUrl: "http://x",
    fetchFn: fetch,
    maxRetries: 2,
    retryBackoffMs: 1,
  });
  const result = await client.authorize({
    tenantId: "t",
    applicationId: "a",
    subject: { type: "user", userId: "u" },
    resource: "x",
    action: "y",
  });
  assert.equal(result.allowed, true);
  assert.equal(attempt, 2);
});

test("4xx surfaces AuthzServiceError without retry", async () => {
  let calls = 0;
  const fetch: typeof globalThis.fetch = async () => {
    calls++;
    return new Response(JSON.stringify({ error: "missing_or_invalid_api_key" }), {
      status: 401,
    });
  };
  const client = new AuthzClient({
    baseUrl: "http://x",
    fetchFn: fetch,
    maxRetries: 3,
    retryBackoffMs: 1,
  });
  await assert.rejects(
    () =>
      client.authorize({
        tenantId: "t",
        applicationId: "a",
        subject: { type: "user", userId: "u" },
        resource: "x",
        action: "y",
      }),
    (err: unknown) =>
      err instanceof AuthzServiceError && err.status === 401,
  );
  assert.equal(calls, 1, "4xx should not be retried");
});

test("ToolGuard.require throws on missing permission", () => {
  const guard = new ToolGuard(["docs.read"]);
  assert.throws(() => guard.require("docs", "write"), PermissionDeniedError);
  assert.doesNotThrow(() => guard.require("docs", "read"));
});

test("MCPGuard reports allowed tools per server", () => {
  const guard = new MCPGuard(["mcp.github.read_repo", "mcp.github.create_issue"]);
  assert.deepEqual(guard.allowedTools("github").sort(), ["create_issue", "read_repo"]);
  assert.equal(guard.isAllowed("slack", "send"), false);
});

test("startAgentSession yields working guards", async () => {
  const { fetch } = makeMockFetch(() => ({
    status: 200,
    body: { permissions: ["docs.read", "mcp.github.read_repo"] },
  }));
  const client = new AuthzClient({ baseUrl: "http://x", fetchFn: fetch });
  const session = await startAgentSession(client, {
    tenantId: "t",
    applicationId: "a",
    userId: "u",
    agentId: "agent",
  });
  assert.ok(session.tools.isAllowed("docs", "read"));
  assert.ok(session.mcp.isAllowed("github", "read_repo"));
  assert.equal(session.mcp.isAllowed("github", "create_issue"), false);
});

File: sdks/go/authz/client_test.go
package authz

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync/atomic"
	"testing"
	"time"
)

func TestAuthorize_AllowAndDeny(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("X-API-Key") != "k1" {
			http.Error(w, "unauthorized", 401)
			return
		}
		if r.URL.Path != "/v1/authorize" {
			http.NotFound(w, r)
			return
		}
		var req AuthorizeRequest
		_ = json.NewDecoder(r.Body).Decode(&req)
		allowed := req.Action == "read"
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(AuthorizeResponse{
			Allowed:            allowed,
			Decision:           map[bool]string{true: "allow", false: "deny"}[allowed],
			Reason:             map[bool]string{true: "permission_granted", false: "missing_permission"}[allowed],
			RequiredPermission: req.Resource + "." + req.Action,
		})
	}))
	defer server.Close()

	client := NewClient(server.URL, "k1")
	ctx := context.Background()

	allow, err := client.Authorize(ctx, AuthorizeRequest{
		TenantID:      "t",
		ApplicationID: "a",
		Subject:       Subject{Type: "user", UserID: "u"},
		Resource:      "docs",
		Action:        "read",
	})
	if err != nil || !allow {
		t.Fatalf("expected allow, got allow=%v err=%v", allow, err)
	}

	allow, err = client.Authorize(ctx, AuthorizeRequest{
		TenantID:      "t",
		ApplicationID: "a",
		Subject:       Subject{Type: "user", UserID: "u"},
		Resource:      "docs",
		Action:        "write",
	})
	if err != nil || allow {
		t.Fatalf("expected deny, got allow=%v err=%v", allow, err)
	}
}

func TestRequire_ReturnsErrPermissionDenied(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewEncoder(w).Encode(AuthorizeResponse{Allowed: false, Decision: "deny"})
	}))
	defer server.Close()
	client := NewClient(server.URL, "")
	err := client.Require(context.Background(), AuthorizeRequest{
		TenantID:      "t",
		ApplicationID: "a",
		Subject:       Subject{Type: "user", UserID: "u"},
		Resource:      "x",
		Action:        "y",
	})
	if !errors.Is(err, ErrPermissionDenied) {
		t.Fatalf("expected ErrPermissionDenied, got %v", err)
	}
}

func TestEffectivePermissions_CachesResult(t *testing.T) {
	var calls int32
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		atomic.AddInt32(&calls, 1)
		_ = json.NewEncoder(w).Encode(map[string][]string{
			"permissions": {"docs.read", "tools.gmail.send"},
		})
	}))
	defer server.Close()

	client := NewClient(server.URL, "", WithCache(60*time.Second, 10))
	ctx := context.Background()

	for i := 0; i < 3; i++ {
		perms, err := client.GetEffectivePermissions(ctx, EffectivePermissionsRequest{
			TenantID: "t", ApplicationID: "a",
			Subject: Subject{Type: "user", UserID: "u"},
		})
		if err != nil {
			t.Fatalf("error: %v", err)
		}
		if _, ok := perms["docs.read"]; !ok {
			t.Fatalf("expected docs.read in result, got %v", perms)
		}
	}
	if got := atomic.LoadInt32(&calls); got != 1 {
		t.Fatalf("expected single upstream call thanks to cache, got %d", got)
	}
}

func TestPost_RetriesOnServerError(t *testing.T) {
	var calls int32
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		n := atomic.AddInt32(&calls, 1)
		if n < 2 {
			http.Error(w, "boom", 500)
			return
		}
		_ = json.NewEncoder(w).Encode(AuthorizeResponse{Allowed: true, Decision: "allow"})
	}))
	defer server.Close()

	client := NewClient(server.URL, "", WithRetry(2, 1*time.Millisecond))
	allow, err := client.Authorize(context.Background(), AuthorizeRequest{
		TenantID: "t", ApplicationID: "a",
		Subject: Subject{Type: "user", UserID: "u"},
		Resource: "x", Action: "y",
	})
	if err != nil || !allow {
		t.Fatalf("expected allow after retry; got allow=%v err=%v", allow, err)
	}
	if got := atomic.LoadInt32(&calls); got != 2 {
		t.Fatalf("expected 2 server calls, got %d", got)
	}
}

func TestServiceError_ContainsStatusAndBody(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, `{"error":"missing_or_invalid_api_key"}`, 401)
	}))
	defer server.Close()
	client := NewClient(server.URL, "")
	_, err := client.Authorize(context.Background(), AuthorizeRequest{})
	var se *ServiceError
	if !errors.As(err, &se) {
		t.Fatalf("expected ServiceError, got %v", err)
	}
	if se.StatusCode != 401 {
		t.Fatalf("expected 401, got %d", se.StatusCode)
	}
	if !strings.Contains(se.Error(), "401") {
		t.Fatalf("expected 401 in error message: %s", se.Error())
	}
}

func TestToolGuard_RequireRejectsMissingPermission(t *testing.T) {
	guard := NewToolGuard(map[string]struct{}{"docs.read": {}})
	if err := guard.Require("docs", "write"); !errors.Is(err, ErrPermissionDenied) {
		t.Fatalf("expected ErrPermissionDenied, got %v", err)
	}
	if err := guard.Require("docs", "read"); err != nil {
		t.Fatalf("expected allow, got %v", err)
	}
}

func TestMCPGuard_NamespacesUnderMCP(t *testing.T) {
	guard := NewMCPGuard(map[string]struct{}{"mcp.github.read_repo": {}})
	if !guard.IsAllowed("github", "read_repo") {
		t.Fatalf("expected mcp.github.read_repo allowed")
	}
	if guard.IsAllowed("github", "create_issue") {
		t.Fatalf("expected mcp.github.create_issue denied")
	}
}
